src/protobuf-net.Reflection/CSharpCodeGenerator.cs

```diff
@@ -709,6 +709,62 @@ protected override void WriteExtension(GeneratorContext ctx, FieldDescriptorProt
             }
         }
 
+        /// <summary>
+        /// Starts a service block
+        /// </summary>
+        protected override void WriteServiceHeader(GeneratorContext ctx, ServiceDescriptorProto service, ref object state)
+        {
+            var name = service.Name;
+            ctx.WriteLine($"{GetAccess(GetAccess(service))} interface I{Escape(name)}").WriteLine("{").Indent();
+        }
+        /// <summary>
+        /// Ends a service block
+        /// </summary>
+        protected override void WriteServiceFooter(GeneratorContext ctx, ServiceDescriptorProto service, ref object state)
+        {
+            ctx.Outdent().WriteLine("}").WriteLine();
+        }
+        /// <summary>
+        /// Write a service method
+        /// </summary>
+        protected override void WriteServiceMethod(GeneratorContext ctx, MethodDescriptorProto method, ref object state)
+        {
+            var outputType = Escape(method.OutputType);
+            var inputType = Escape(method.InputType);
+
+            var target = ctx.TryFind<DescriptorProto>(method.OutputType);
+            if (target != null)
+            {
+                var declaringType = target.Parent;
+
+                if (declaringType is IType type)
+                {
+                    var name = FindNameFromCommonAncestor(type, target, ctx.NameNormalizer);
+                    if (!string.IsNullOrWhiteSpace(name))
+                    {
+                        outputType = name;
+                    }
+                }
+            }
+
+            target = ctx.TryFind<DescriptorProto>(method.InputType);
+            if (target != null)
```

Comment:
nit: it feels like this is pure duplication - maybe a local function would be appropriate here?